// prisma-kysely/dist/helpers/generateImplicitManyToManyModels.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const generateImplicitManyToManyModels_1 = require("./generateImplicitManyToManyModels");
(0, vitest_1.test)("it respects overrides when generating field types", () => {
    const newModels = (0, generateImplicitManyToManyModels_1.generateImplicitManyToManyModels)([
        {
            name: "Category",
            fields: [
                {
                    name: "id",
                    type: "String",
                    isId: true,
                    hasDefaultValue: true,
                    isList: false,
                    isReadOnly: false,
                    isRequired: true,
                    isUnique: true,
                    kind: "scalar",
                },
                {
                    name: "posts",
                    kind: "object",
                    isList: true,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    type: "Post",
                    hasDefaultValue: false,
                    relationName: "CategoryToPost",
                    relationFromFields: [],
                    relationToFields: [],
                    isGenerated: false,
                    isUpdatedAt: false,
                },
            ],
            primaryKey: null,
            dbName: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
        {
            name: "Post",
            fields: [
                {
                    name: "id",
                    type: "String",
                    isId: true,
                    hasDefaultValue: true,
                    isList: false,
                    isReadOnly: false,
                    isRequired: true,
                    isUnique: true,
                    kind: "scalar",
                },
                {
                    name: "categories",
                    kind: "object",
                    isList: true,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    type: "Category",
                    hasDefaultValue: false,
                    relationName: "CategoryToPost",
                    relationFromFields: [],
                    relationToFields: [],
                    isGenerated: false,
                    isUpdatedAt: false,
                },
            ],
            primaryKey: null,
            dbName: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
    ]);
    (0, vitest_1.expect)(newModels).toEqual([
        {
            name: "CategoryToPost",
            dbName: "_CategoryToPost",
            fields: [
                {
                    hasDefaultValue: false,
                    isId: false,
                    isList: false,
                    isReadOnly: true,
                    isRequired: true,
                    isUnique: false,
                    kind: "scalar",
                    name: "A",
                    type: "String",
                },
                {
                    hasDefaultValue: false,
                    isId: false,
                    isList: false,
                    isReadOnly: true,
                    isRequired: true,
                    isUnique: false,
                    kind: "scalar",
                    name: "B",
                    type: "String",
                },
            ],
            primaryKey: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
    ]);
});
(0, vitest_1.test)("it filters out many-to-one relations safely", () => {
    const newModels = (0, generateImplicitManyToManyModels_1.generateImplicitManyToManyModels)([
        {
            name: "Category",
            fields: [
                {
                    name: "id",
                    type: "String",
                    isId: true,
                    hasDefaultValue: true,
                    isList: false,
                    isReadOnly: false,
                    isRequired: true,
                    isUnique: true,
                    kind: "scalar",
                },
                {
                    name: "posts",
                    kind: "object",
                    isList: true,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    type: "Post",
                    hasDefaultValue: false,
                    relationName: "CategoryToPost",
                    relationFromFields: [],
                    relationToFields: [],
                    isGenerated: false,
                    isUpdatedAt: false,
                },
            ],
            primaryKey: null,
            dbName: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
        {
            name: "Post",
            fields: [
                {
                    name: "id",
                    type: "String",
                    isId: true,
                    hasDefaultValue: true,
                    isList: false,
                    isReadOnly: false,
                    isRequired: true,
                    isUnique: true,
                    kind: "scalar",
                },
                {
                    name: "categories",
                    kind: "object",
                    isList: true,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    type: "Category",
                    hasDefaultValue: false,
                    relationName: "CategoryToPost",
                    relationFromFields: [],
                    relationToFields: [],
                    isGenerated: false,
                    isUpdatedAt: false,
                },
                {
                    name: "userId",
                    kind: "scalar",
                    isList: false,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: true,
                    hasDefaultValue: false,
                    type: "String",
                    isGenerated: false,
                    isUpdatedAt: false,
                },
                {
                    name: "user",
                    kind: "object",
                    isList: false,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    hasDefaultValue: false,
                    type: "User",
                    relationName: "PostToUser",
                    relationFromFields: ["userId"],
                    relationToFields: ["id"],
                    relationOnDelete: "Cascade",
                    isGenerated: false,
                    isUpdatedAt: false,
                },
            ],
            primaryKey: null,
            dbName: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
        {
            name: "User",
            fields: [
                {
                    name: "id",
                    type: "String",
                    isId: true,
                    hasDefaultValue: true,
                    isList: false,
                    isReadOnly: false,
                    isRequired: true,
                    isUnique: true,
                    kind: "scalar",
                },
                {
                    name: "posts",
                    kind: "object",
                    isList: true,
                    isRequired: true,
                    isUnique: false,
                    isId: false,
                    isReadOnly: false,
                    type: "Post",
                    hasDefaultValue: false,
                    relationName: "PostToUser",
                    relationFromFields: [],
                    relationToFields: [],
                    isGenerated: false,
                    isUpdatedAt: false,
                },
            ],
            primaryKey: null,
            dbName: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
    ]);
    (0, vitest_1.expect)(newModels).toEqual([
        {
            name: "CategoryToPost",
            dbName: "_CategoryToPost",
            fields: [
                {
                    hasDefaultValue: false,
                    isId: false,
                    isList: false,
                    isReadOnly: true,
                    isRequired: true,
                    isUnique: false,
                    kind: "scalar",
                    name: "A",
                    type: "String",
                },
                {
                    hasDefaultValue: false,
                    isId: false,
                    isList: false,
                    isReadOnly: true,
                    isRequired: true,
                    isUnique: false,
                    kind: "scalar",
                    name: "B",
                    type: "String",
                },
            ],
            primaryKey: null,
            uniqueFields: [],
            uniqueIndexes: [],
        },
    ]);
});
